### PR TITLE
Add OpenTelemetry Sample

### PR DESCRIPTION
**Pull request by robcao**
## What was changed

Added OpenTelemetry sample.

## Why?

It was missing and requested.

## Checklist

1. Closes 
https://github.com/temporalio/samples-dotnet/issues/67

3. How was this tested:

Added screenshots similar to the python sample

4. Any docs updates needed?

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -19,20 +19,20 @@
...
     Can also reference the SDK downloaded to a local directory:
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\temporal-sdk-dotnet\src\Temporalio\Temporalio.csproj" />
     -->
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
-    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
-    <PackageReference Include="Microsoft.VisualStudio.Threading.Analyzers" Version="17.4.33" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
```
Comment (robcao) [inline Directory.Build.props:33]: Had to update logging here because the OpenTelemetry 1.9.0+ library has a dependency on 8.0.0+

image

File: src/OpenTelemetry/Program.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+using Temporalio.Extensions.OpenTelemetry;
+using Temporalio.Runtime;
+using Temporalio.Worker;
+using TemporalioSamples.OpenTelemetry;
+
+AssemblyName assemblyName = typeof(TemporalClient).Assembly.GetName();
+
+using var meter = new Meter(assemblyName.Name!, assemblyName.Version!.ToString());
+
+string instanceId = args.ElementAtOrDefault(0) ?? throw new ArgumentException("Must pass 'worker' or 'workflow' as the single argument");
```
Comment (robcao) [inline src/OpenTelemetry/Program.cs]: A little bit of repetition here, but it matters greatly for the OpenTelemetry sample in terms of correlating telemetry coming from the worker or client.

File: src/OpenTelemetry/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+services:
+  aspire-dashboard:
+    environment:
+      Dashboard__Frontend__AuthMode: Unsecured
+    image: mcr.microsoft.com/dotnet/aspire-dashboard:8.0
```
Comment (robcao) [inline src/OpenTelemetry/docker-compose.yaml:5]: I did see that the python samples are using the Jaeger image and not the Aspire Dashboard, which is developed by Microsoft: https://github.com/temporalio/samples-python/tree/main/open_telemetry

I can change to the Jaeger image if desired.

The main benefit of using the Aspire dashboard over Jaeger is that it provides visualizations for metrics, which iirc Jaeger does not. I think using a Microsoft image is _probably_ okay specifically for .NET

File: src/OpenTelemetry/MyWorkflow.workflow.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using Microsoft.Extensions.Logging;
+using Temporalio.Workflows;
+
+[Workflow]
+public class MyWorkflow
+{
+    [WorkflowRun]
+    public async Task<string> RunAsync()
+    {
+        Workflow.Logger.LogInformation("Running workflow {WorkflowId}.", Temporalio.Workflows.Workflow.Info.WorkflowId);
```
Comment (cretz) [inline src/OpenTelemetry/MyWorkflow.workflow.cs:12]:
```suggestion
        Workflow.Logger.LogInformation("Running workflow {WorkflowId}.", Workflow.Info.WorkflowId);
```

File: src/OpenTelemetry/Program.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            AddSimpleConsole(options => options.TimestampFormat = "[HH:mm:ss] ").
+            SetMinimumLevel(LogLevel.Information)),
+    Interceptors = new[] { new TracingInterceptor() },
+    Runtime = new TemporalRuntime(new TemporalRuntimeOptions()
+    {
+        Telemetry = new TelemetryOptions()
+        {
+            Metrics = new MetricsOptions()
+            {
+                CustomMetricMeter = new CustomMetricMeter(meter),
```
Comment (cretz) [inline src/OpenTelemetry/Program.cs:53]: Hrmm, so there are two ways to configure OTel telemetry metrics. One way is this, but the other way is simply to set the OTel options in this metric options so our internal code forwards. We often encourage the latter because it's less setup and it matches other SDKs.

I wonder if there is some way we can show both? Like maybe switched from one to the other via CLI flag?

File: src/OpenTelemetry/Program.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+using Temporalio.Client;
+using Temporalio.Extensions.DiagnosticSource;
+using Temporalio.Extensions.OpenTelemetry;
+using Temporalio.Runtime;
+using Temporalio.Worker;
+using TemporalioSamples.OpenTelemetry;
+
+AssemblyName assemblyName = typeof(TemporalClient).Assembly.GetName();
```
Comment (cretz) [inline src/OpenTelemetry/Program.cs:15]:
```suggestion
var assemblyName = typeof(TemporalClient).Assembly.GetName();
```

Feel free to use `var` throughout this file

File: src/OpenTelemetry/Properties/launchSettings.json
Comment (cretz) [inline src/OpenTelemetry/Properties/launchSettings.json:1]: Is this file needed? We don't usually add to the rest of our samples

File: src/OpenTelemetry/docker-compose.yaml (hunk 3 above)
Comment (cretz) [reply src/OpenTelemetry/docker-compose.yaml:5]: I don't mind this, I like the metrics. Alternatively, you can get rid of docker compose and just tell people how to start the this docker container directly in the README, but this is fine too.

File: src/OpenTelemetry/Activities.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.Extensions.Logging;
+using Temporalio.Activities;
+
+namespace TemporalioSamples.OpenTelemetry
```
Comment (cretz) [inline src/OpenTelemetry/Activities.cs:4]: Use the top-of-file namespace declaration style like is done in the workflow class

**Review by cretz — COMMENTED** [review]
This is great, thanks! Just some minor things, but overall LGTM

File: src/OpenTelemetry/docker-compose.yaml (hunk 3 above)
Comment (robcao) [reply src/OpenTelemetry/docker-compose.yaml:5]: Imo, docker compose is easier with a single command, since environment variables and port forwarding are evolved

File: src/OpenTelemetry/Activities.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public static class Activities
+    {
+        [Activity]
+        public static void MyActivity(string input)
```
[the remark this answers is not in the document]
Comment (robcao) [reply src/OpenTelemetry/Activities.cs]: Added

File: src/OpenTelemetry/Properties/launchSettings.json
Comment (robcao) [reply src/OpenTelemetry/Properties/launchSettings.json:1]: Not needed, removed.

launchSettings are a convenience that popular .NET IDEs such as Rider, VS, and VSCode support that allow developers to pre-configure executables launched directly in the IDE with command line arguments, environment variables, etc

File: src/OpenTelemetry/Program.cs (hunk 5 above)
Comment (robcao) [reply src/OpenTelemetry/Program.cs]: I split this into two samples,  `OpenTelemetry/CoreSdkFowarding` and `OpenTelemetry/DotNetMetrics`.

I thought this was cleanest, and follows a similar pattern that exists with the polling samples

Comment (cretz) [reply src/OpenTelemetry/Program.cs]: Ok, will look. Alternatively we probably could have just had a different parameter accepted in the `Program.cs` to choose one approach or the other.

File: src/OpenTelemetry/CoreSdkForwarding/TemporalioSamples.OpenTelemetry.CoreSdkForwarding.csproj [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
Comment (cretz) [inline src/OpenTelemetry/CoreSdkForwarding/TemporalioSamples.OpenTelemetry.CoreSdkForwarding.csproj:7]:
```suggestion
```

These are already in `Directory.Build.props`, we want to keep them centralized (same for other csproj in this PR)

File: src/OpenTelemetry/CoreSdkForwarding/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# OpenTelemetry - .NET Metrics
+
+This sample shows how to configure the SDK to forward metrics from the Core SDK.
+
+The main advantage over using .NET metrics is simplicity; additionally, there is no need to take a dependency on the .NET OpenTelemetry libraries.
+
+Note that in order to set up tracing, .NET OpenTelemetry must be used. See the [DotNetMetrics](../DotNetMetrics) sample for an example of how to set up tracing.
```
Comment (cretz) [inline src/OpenTelemetry/CoreSdkForwarding/README.md:7]: This is a bit confusing. This sample should probably add OTel tracing interceptor too. Many of our users setup OTel interceptor tracing w/ meter and such and still using the Core SDK-based OTel metrics instead of the .NET custom meter one.

File: README.md [hunk 11]
```diff
@@ -21,6 +21,7 @@ Prerequisites:
 * [DependencyInjection](src/DependencyInjection) - How to inject dependencies in activities and use generic hosts for workers
 * [Encryption](src/Encryption) - End-to-end encryption with Temporal payload codecs.
 * [Mutex](src/Mutex) - How to implement a mutex as a workflow. Demonstrates how to avoid race conditions or parallel mutually exclusive operations on the same resource.
+* [OpenTelemetry](src/OpenTelemetry) - Demonstrates how to set up OpenTelemetry tracing and metrics for both the client and worker, using both the .NET metrics API and internal forwarding from the Core SDK.
```
Comment (cretz) [inline README.md:24]: If you are going to keep these as separate projects, may want a `src/OpenTelemetry/README.md` that shows the two subdir projects (granted I don't think we need entirely separate projects for something as simple as switching the metric config).

**Review by cretz — COMMENTED** [review]
Overall LGTM. Would like to have OTel tracing in both and I really, I think it's clearer to have just one project w/ metric config switched based on command line (if you can't choose which to default, make the CLI arg required to pick one, e.g. `--dotnet-metrics` and `--core-metrics` or something).

**Comment by cretz** [comment]
@robcao - assuming you don't mind, I am going to work in your branch to get this over the finish line. I think it may just be a matter of merging main (fixing conflicts) and a few extra things.

File: src/OpenTelemetry/CoreSdkForwarding/TemporalioSamples.OpenTelemetry.CoreSdkForwarding.csproj (hunk 9 above)
Comment (cretz) [reply src/OpenTelemetry/CoreSdkForwarding/TemporalioSamples.OpenTelemetry.CoreSdkForwarding.csproj]: Fixed

File: src/OpenTelemetry/CoreSdkForwarding/README.md (hunk 10 above)
Comment (cretz) [reply src/OpenTelemetry/CoreSdkForwarding/README.md]: Fixed

File: src/OpenTelemetry/CoreSdkForwarding/Program.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
+﻿using System.Diagnostics.Metrics;
+using Microsoft.Extensions.Logging;
+using Temporalio.Client;
+using Temporalio.Runtime;
+using Temporalio.Worker;
+using TemporalioSamples.OpenTelemetry.Common;
+
+var assemblyName = typeof(TemporalClient).Assembly.GetName();
+
+using var meter = new Meter(assemblyName.Name!, assemblyName.Version!.ToString());
```
[the remark this answers is not in the document]
Comment (cretz) [reply src/OpenTelemetry/CoreSdkForwarding/Program.cs]: Fixed

File: README.md (hunk 11 above)
Comment (cretz) [reply README.md:26]: Fixed

**Comment by robcao** [comment]
> @robcao - assuming you don't mind, I am going to work in your branch to get this over the finish line. I think it may just be a matter of merging main (fixing conflicts) and a few extra things.

Yeah, I don't mind.

Sorry about the unintentional abandonment of the pull request, I had meant to get back to it, but life just got in the way.

I don't mind collapsing the samples into a single project file. My original thought was that the two different techniques used to gather metrics have different dependencies, so that putting them together in the same project file makes it harder to understand the minimal set of dependencies, but I am ok either way.

**Comment by cretz** [comment]
All good, I think we're ready for merge, thanks for the contribution!

**Review by cretz — APPROVED** [review]
